### ISSUE TITLE
How to return Error response?

### ISSUE THREAD

**Issue by idangozlan** [malijs/mali#24]
Is there any way return Error response to client? can't find it on the docs.

**Comment by bojand** [comment]
Hello, in `UNARY` and `REQUEST_STREAM` calls you can just throw the error and it will be sent to the client. There is some documentation in the section [on error handling](https://malijs.github.io/#/errors) as well as examples [in tests](https://github.com/malijs/mali/blob/master/test/errors.test.js#L45]. In stream responses that depends on the behaviour you desire. You can either emit an error, or probably more appropriately have error data within the payload and send those along within the stream, and check on client side within the `'data'` event handler. I will try and improve documentation regarding errors and provide some more examples.

**Comment by idangozlan** [comment]
Thanks.
It's not clear to me why the server has to handle error while the expected
behavior is to send error to the client (the error throwing is appear on
the server), instead of that (and as the original grpc server behavior) it
should allow just send the error without triggering it on the server too..

On Jan 17, 2018 19:14, "Bojan" <notifications@github.com> wrote:

Hello, in UNARY and REQUEST_STREAM calls you can just throw the error and
it will be sent to the client. There is some documentation in the section on
error handling <https://malijs.github.io/#/errors> as well as examples [in
tests](https://github.com/malijs/mali/blob/master/test/errors.test.js#L45].
In stream responses that depends on the behaviour you desire. You can
either emit an error, or probably more appropriately have error data within
the payload and send those along within the stream, and check on client
side within the 'data' event handler. I will try and improve documentation
regarding errors and provide some more examples.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/malijs/mali/issues/24#issuecomment-358524888>, or mute
the thread

**Comment by bojand** [comment]
Mali is built on top of the core grpc Node module and utilizes promise based API. With `async` / `await` the main mechanism for errors is the standard `try` / `catch`. In case of an error we catch it and send it to the client. The error logging can be suppressed by setting the `app.silent` settings to `true`. 

I have just added the ability to explicitly set the response to an error, and if so, we will send it to the client circumventing the application error logging regardless of the `app.silent` setting.

```js
async function hello (ctx) {
  ctx.res = new Error('boom!')
}
```

I have also added additional [documentation](https://malijs.github.io/#/errors) regarding error handling and communication, as well as corresponding [example code](https://github.com/malijs/examples).

Hope this helps.
